File: Part 2 - Regression/Multiple Linear Regression/multiple_linear_regression.R
#Created on Sun Mar 29 2020
#@author: Swapnil Vivek Kulkarni

#Multiple Linear regression

#import dataset
dataset = read.csv('50_Startups.csv')

#Encoding categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3)) #C() is nothing but vector in R
#Spliting dataset into Training set and Test set 
#install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit,SplitRatio = 0.8)
training_set = subset(dataset,split == TRUE)
test_set = subset(dataset, split == FALSE)

#Fitting multiple linear regresson in to Training set
regressor = lm(formula = Profit ~ .,
               data = training_set) #. means all the independeant columns

#Predict testset results
y_pred = predict(regressor,newdata = test_set)

#Building optimal model usingbackword elimination
regressor = lm(formula =  Profit ~ R.D.Spend + Administration + Marketing.Spend + State,
              data = dataset)
summary(regressor)

regressor = lm(formula =  Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor)

regressor = lm(formula =  Profit ~ R.D.Spend,
               data = dataset)
summary(regressor)

File: Part 2 - Regression/Multiple Linear Regression/multiple_linear_regression.py
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 25 2020

@author: Swapnil Kulkarni
"""
#Multiple Linear regression

#Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Imporing the dataset 
dataset = pd.read_csv('50_Startups.csv') 
X = dataset.iloc[:,:-1].values # : means all elements
Y = dataset.iloc[:,4].values

#Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,3] = labelencoder_X.fit_transform(X[:,3])
onehotencoder = OneHotEncoder(categorical_features=[3])
X = onehotencoder.fit_transform(X).toarray()

#Avoiding Dummy variable trap
X = X[:,1:] #Always omit one dummy variable

#splitting the dataset into training set and test set
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=0)

#Fitting multiple linear regression into traing set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, Y_train)

#Predecting the test set results
Y_pred = regressor.predict(X_test)
    
#Building optimal model using backword elemination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values=X, axis = 1)
X_opt = X[:,[0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:,[0,1,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:,[0,3,4,5]]
regressor_OLS = sm.OLS(endog=Y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:,[0,3,5]]
regressor_OLS = sm.OLS(endog=Y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:,[0,3]]
regressor_OLS = sm.OLS(endog=Y, exog=X_opt).fit()
regressor_OLS.summary()